src/viam/sdk/components/board.hpp

```diff
@@ -221,6 +232,22 @@ class Board : public Component {
     virtual digital_value read_digital_interrupt(const std::string& digital_interrupt_name,
                                                  const AttributeMap& extra) = 0;
 
+    /// @brief Returns a stream of digital interrupt ticks.
+    /// @param digital_interrupt_names digital interrupts to stream
+    /// @param ticks queue to put the ticks in
+    inline void stream_ticks(std::vector<std::string> const& digital_interrupt_names,
+                             std::shared_ptr<std::queue<tick>> ticks) {
+        return stream_ticks(digital_interrupt_names, ticks, {});
+    }
+
+    /// @brief Returns a stream of digital interrupt ticks.
+    /// @param digital_interrupt_names digital interrupts to stream
+    /// @param extra Any additional arguments to the method
+    /// @param ticks queue to put the ticks in
+    virtual void stream_ticks(std::vector<std::string> const& digital_interrupt_names,
```

Comment:
Now that I understand a little more about the intention here, I think a change of design is likely needed.

My understanding, and please let me know if I have this wrong, is that the caller of `stream_ticks` should not return until the request is cancelled (how?) and that all ticks that are returned during that period should be available to act on _as they occur_.

However, this interface cannot safely be made to do that. The calling thread is blocked, so some other thread must be reacting to the arriving events. To do so, they would need to be consuming those events from the same queue that the thread which called `stream_ticks` is producing the ticks into.

But the queue passed here has no means to synchronize, nor does it have a way to communicate wakeups from the producer (the thread blocked in the streaming call) to the consumer (the thread that wants to pull events from the queue).

Otherwise, with the API here, the only safe thing to do is to:
1) Have one thread call `stream_ticks`, it blocks
2) Someone else cancels the call (how?)
3) The call to `stream_ticks` unblocks, and now you have a queue with all of the ticks that happened.

But I'm fairly sure that isn't what is desired.

I see two ways forward. The first would be to modify this such that the argument to `stream_ticks` is a self-synchronized single-consumer single-producer queue, probably over `Optional<tick>` so there is a way to indicate completion:

```
pc_queue<Optional<Tick>> tick_queue;
std::thread consumer([&tick_queue] {
    auto next = tick_queue.wait();
    if (!next) break;
    std::cout << "Saw a tick: " << ... ;
});
auto board_client = ...
board_client->stream_ticks(pc_queue);
consumer.join();
```

The other way would be to have `stream_ticks` take an on-tick callback.

```
virtual void stream_ticks(std::function<bool(Tick tick)>& tick_handler) = 0;
```

Then you could use it like this:

```
auto board_client = ...
board_client->stream_ticks([](Tick tick) {
    std::cout << "Saw a tick: " << ...;
    if (...) { // Don't want more ticks
       return false;
    return true;
});
```

Where the callback and return a value indicating that it wants no more ticks. Here, the thread doing the streaming also invokes the callback, so some care needs to be taken not to block. The callback, for instance, could then decide to push the ticks to a queue that some other thread is consuming from:

```
pc_queue<Tick> tick_queue;
std::thread consumer([&tick_queue] {
    auto next = tick_queue.wait();
    std::cout << "Saw a tick: " << ... ;
});

auto board_client = ...
board_client->stream_ticks([&queue](Tick tick) {
    queue.push(std::move(tick));
});
consumer.join();
```

There are probably other ways. I'd like to get @stuqdog's thoughts on the matter too so we can design the right API.